xarray/core/computation.py

```diff
@@ -1151,6 +1151,99 @@ def earth_mover_distance(first_samples,
         return apply_array_ufunc(func, *args, dask=dask)
 
 
+def call_on_dataset(func, obj, *args, **kwargs):
+    """apply a function expecting a Dataset to a xarray object
+
+    Parameters
+    ----------
+    func : callable
+        A function expecting a Dataset as its first parameter.
+    obj : DataArray or Dataset
+        The dataset to apply ``func`` to. If a ``DataArray``, convert it to a single
+        variable ``Dataset`` first.
+    *args, **kwargs
+        Additional arguments to ``func``
+
+    Returns
+    -------
+    DataArray or Dataset
+        The result of ``func(obj, *args, **kwargs)`` with the same type as ``obj``.
+
+    Notes
+    -----
+    DataArray objects without a name (or named ``None``) will be renamed to
+    ``"<this-array>"`` before being passed to ``func``. The empty name will be restored
+    for the result of the call.
+
+    See Also
+    --------
+    Dataset.map
+    Dataset.pipe
+    DataArray.pipe
+
+    Examples
+    --------
+    >>> def f(ds):
+    ...     return xr.Dataset(
+    ...         {
+    ...             name: var * var.attrs.get("scale", 1)
+    ...             for name, var in ds.data_vars.items()
+    ...         },
+    ...         coords=ds.coords,
+    ...         attrs=ds.attrs,
+    ...     )
+    ...
+    >>> ds = xr.Dataset(
+    ...     {"a": ("x", [3, 4], {"scale": 0.5}), "b": ("x", [-1, 1], {"scale": 1.5})},
+    ...     coords={"x": [0, 1]},
+    ...     attrs={"attr": "value"},
+    ... )
+    >>> ds
+    <xarray.Dataset>
+    Dimensions:  (x: 2)
+    Coordinates:
+      * x        (x) int64 0 1
+    Data variables:
+        a        (x) int64 3 4
+        b        (x) int64 -1 1
+    Attributes:
+        attr:     value
+    >>> xr.call_on_dataset(f, ds)
+    <xarray.Dataset>
+    Dimensions:  (x: 2)
+    Coordinates:
+      * x        (x) int64 0 1
+    Data variables:
+        a        (x) float64 1.5 2.0
+        b        (x) float64 -1.5 1.5
+    Attributes:
+        attr:     value
+    >>> xr.call_on_dataset(f, ds.a)
+    <xarray.DataArray 'a' (x: 2)>
+    array([1.5, 2. ])
+    Coordinates:
+      * x        (x) int64 0 1
+    """
+    from .dataarray import _THIS_ARRAY, DataArray
+    from .parallel import dataarray_to_dataset, dataset_to_dataarray
+
+    if isinstance(obj, DataArray):
+        ds = dataarray_to_dataset(obj)
+        if obj.name is None:
+            ds = ds.rename({_THIS_ARRAY: "<this-array>"})
```

Comment:
another option would be to use `backends.core.api.DATAARRAY_VARIABLE` which is used when writing a DataArray to netcdf (I think). I don't feel strongly about this.

Reply:
yeah, I don't know which name would be better, either. `THIS_ARRAY` is not part of the public API so we can't use it, and `None` obviously doesn't work, either. Using a string seems like a good choice but the exact value will almost always be arbitrary. The advantage of `"<this-array>"` is that it is the string representation of `THIS_ARRAY`, but that's the only reason I chose that. `DATAARRAY_VARIABLE` or `DATAARRAY_NAME` have the value `"__xarray_dataarray_name__"`, but neither of them are actually part of the public API (I think?), which means they have the same issue as `THIS_ARRAY` (not sure if that's actually a problem, though: the simply reference a string).

Reply:
I thought @keewis 's idea of `self` was good from https://github.com/pydata/xarray/pull/5493#discussion_r654698258, to the extent that could apply here

Reply:
I used this in xarray-contrib/pint-xarray#110, and for that at least it's actually an advantage that I have to pass the name. Not sure if that's the same for every other use case though (but defining the name explicitly is not much overhead so it should be fine)